### PR TITLE
Use `ENT_QUOTES|ENT_SUBSTITUTE|ENT_HTML5` instead of just `ENT_QUOTES`

### LINKED ISSUE #7245: Wrong single quote encoding on email (&apos;)
### Affected version(s)

4.13.40 -> 4.13.44

### Description

When sending an email from a form generated by Contao, the single quote are badly encoded and look like this: `&apos;`
I was unable to test versions prior to 4.13.40

### Possible solution

add `ENT_QUOTES|ENT_HTML5` argument to `StringUtil::decodeEntities` on `core-bundle/src/Resources/contao/forms/Form.php:479`

`$email->text = StringUtil::decodeEntities(trim($message), ENT_QUOTES|ENT_HTML5) . $uploaded . "\n\n";`

### PR DESCRIPTION
**Pull request by leofeyer**
Fixes #7245

@ausi What about the function calls that use the default arguments?

https://github.com/contao/contao/blob/0a6e74ab81a0debe7594b1f4d787960ebbc329db/core-bundle/src/Controller/CaptchaController.php#L50

https://github.com/contao/contao/blob/0a6e74ab81a0debe7594b1f4d787960ebbc329db/core-bundle/src/Resources/contao/drivers/DC_Folder.php#L660

The default is `ENT_QUOTES | ENT_SUBSTITUTE` IIRC.

### CODE REVIEW

**Comment by ausi** [comment]
> What about the function calls that use the default arguments?

We should use `ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML5` there.

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php [hunk 1]
```diff
@@ -219,7 +219,7 @@ public static function substrHtml($strString, $intNumberOfChars)
 	 *
 	 * @return string The decoded string
 	 */
-	public static function decodeEntities($strString, $strQuoteStyle=ENT_QUOTES, $strCharset=null)
+	public static function decodeEntities($strString, $strQuoteStyle=ENT_QUOTES|ENT_HTML5, $strCharset=null)
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: I think we should keep the default here and use `$strQuoteStyle|ENT_HTML5` below.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: Then we could not pass an argument without `| ENT_HTML5` anymore.

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: That should be fine I think?

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: I don’t see a use case for using `html_entity_decode()` without `ENT_HTML5` (encode can be different, but not decode IMO).

**Comment by leofeyer** [comment]
> We should use `ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML5` there.

Like this? f8879ecf6434faa2282cb2391880c5c4c08e5066

It seems weird to me to sometimes use `ENT_SUBSTITUTE ` and sometimes not. Should we not use `ENT_QUOTES | ENT_HTML5` consistently?

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: Maybe there is one, or maybe not. But if we hardcode `| ENT_HTML5`, people have no option to override it anymore. Therefore, I would rather only adjust the default value.

**Comment by ausi** [comment]
> Should we not use `ENT_QUOTES | ENT_HTML5` consistently?

Yes we should use the same everywhere, but it needs to be `ENT_QUOTES | ENT_SUBSTITUTE | ENT_HTML5` because of [from the docs](https://www.php.net/manual/en/function.htmlspecialchars.php#refsect1-function.htmlspecialchars-returnvalues): 
> If the input string contains an invalid code unit sequence within the given encoding an empty string will be returned, unless either the ENT_IGNORE or ENT_SUBSTITUTE flags are set.

File: core-bundle/src/Resources/contao/library/Contao/StringUtil.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: I’d still prefer adding `| ENT_SUBSTITUTE | ENT_HTML5` below and only use the `$strQuoteStyle` parameter for selecting the desired quote-style (ENT_COMPAT, ENT_QUOTES, ENT_NOQUOTES).
But I’m fine either way.

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/StringUtil.php]: Changed in 0cd7f956136e216be7db5e306145a86141f25756.

**Review by ausi — APPROVED** [review]

**Pull request merged**